### PR TITLE
feat: add alerting to e2e module

### PR DESCRIPTION
**Pull request by sailorlqh**
fix: #174

### CODE REVIEW

File: terraform/modules/monitoring/alert.tf [hunk 1]
```diff
@@ -112,11 +112,11 @@ resource "google_monitoring_alert_policy" "pmap_messages_too_old" {
 }
 
 resource "google_monitoring_alert_policy" "pmap_num_of_undeliverable_messages_oversized" {
-  for_each = toset(var.pmap_subscription_ids)
+  for_each = var.pmap_subscription_ids
 
   project = var.project_id
 
-  display_name = "Nnumber of undeliverable messages for pmap subscription id {${each.key}} oversized[MEAN]"
+  display_name = "Nnumber of undeliverable messages for pmap subscription id {${each.value}} oversized[MEAN]"
```
Comment (yolocs) [inline terraform/modules/monitoring/alert.tf:119]:
```suggestion
  display_name = "Number of undeliverable messages for pmap subscription id {${each.value}} oversized[MEAN]"
```

File: terraform/e2e/main.tf [hunk 2]
```diff
@@ -15,6 +15,14 @@
 locals {
   mapping_service_name = "mapping"
   policy_service_name  = "policy"
+
+  alert_subscription_map = {
+    mapping_bigquery         = module.common_infra.bigquery_subscriptions[local.mapping_service_name].event_subscription_id,
+    mapping_bigquery_failure = module.common_infra.bigquery_subscriptions[local.mapping_service_name].failure_subscription_id,
+    policy_bigquery          = module.common_infra.bigquery_subscriptions[local.policy_service_name].failure_subscription_id,
```
Comment (capri-xiyue) [inline terraform/e2e/main.tf:22]: Should we have both policy_bigquery and policy_bigquery_failure?

File: terraform/modules/monitoring/alert.tf [hunk 3]
```diff
@@ -78,11 +78,11 @@ resource "google_monitoring_alert_policy" "prober_service_success_number_below_t
 }
 
 resource "google_monitoring_alert_policy" "pmap_messages_too_old" {
-  for_each = toset(var.pmap_subscription_ids)
+  for_each = var.pmap_subscription_ids
```
Comment (capri-xiyue) [inline terraform/modules/monitoring/alert.tf:81]: Why change this?

File: terraform/modules/monitoring/variables.tf [hunk 4]
```diff
@@ -102,8 +102,8 @@ variable "log_level" {
 }
 
 variable "pmap_subscription_ids" {
-  type        = list(string)
-  default     = []
+  type        = map(string)
```
Comment (capri-xiyue) [inline terraform/modules/monitoring/variables.tf:105]: I think a list will be more user friendly.

File: terraform/e2e/main.tf [hunk 5]
```diff
@@ -15,6 +15,14 @@
 locals {
   mapping_service_name = "mapping"
   policy_service_name  = "policy"
+
+  alert_subscription_map = {
```
Comment (capri-xiyue) [inline terraform/e2e/main.tf:19]: Why use map here?

Comment (capri-xiyue) [inline terraform/e2e/main.tf:19]: If you want to introduce more readability in the e2e module, it is ok that you can local map here. But you should not enforce all users to have a map when using monitoring module.

File: terraform/modules/monitoring/variables.tf (hunk 4 above)
Comment (sailorlqh) [reply terraform/modules/monitoring/variables.tf:105]: Terraform will complain and throw out errors if it’s a list and says it prefers a map. I have tried using a list, it didn’t work, so I changed to this.

Comment (sailorlqh) [reply terraform/modules/monitoring/variables.tf:105]: If map isn’t wanted, I can try if set works.

Comment (capri-xiyue) [reply terraform/modules/monitoring/variables.tf:105]: I ran terraform apply and it worked without any warning.

Comment (capri-xiyue) [reply terraform/modules/monitoring/variables.tf:105]: Discussed offline, the original module should work as I changed the list to set in the monitoring module.

Comment (sailorlqh) [reply terraform/modules/monitoring/variables.tf]: Updated.

File: terraform/e2e/main.tf (hunk 2 above)
Comment (sailorlqh) [reply terraform/e2e/main.tf]: Added

File: terraform/e2e/main.tf (hunk 5 above)
Comment (sailorlqh) [reply terraform/e2e/main.tf]: changed to list.

**Review by yolocs — APPROVED** [review]
lgtm. Will let Xiyue double check

File: terraform/modules/monitoring/variables.tf (hunk 4 above)
Comment (sailorlqh) [reply terraform/modules/monitoring/variables.tf:113]: Discussed offline, reproduced the real error, and decided to use map and some comments to describe why we used map here.

File: terraform/e2e/main.tf (hunk 2 above)
Comment (capri-xiyue) [reply terraform/e2e/main.tf:22]: Did you add it?

**Review by capri-xiyue — APPROVED** [review]

File: terraform/e2e/main.tf (hunk 2 above)
Comment (sailorlqh) [reply terraform/e2e/main.tf]: sorry it's somehow override by other changes, added now.

**Pull request merged**